File: 101-200/161-170/162-findPeakElement/findPeakElement.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int left = 0, right = nums.size();
        int edgeValue = INT_MIN;

        while (left < right) {
            int middle = (left + right) / 2;

            if (nums[middle] > (middle + 1 == nums.size() ? edgeValue : nums[middle+1]))
                left = middle + 1;
            else if (nums[middle] > (middle ? nums[middle-1] : edgeValue))
                right = miidle - 1;
            else
                return middle;
        }

        return -1;
    }

    int findPeakElement2(vector<int> &num)
    {
        int low = 0;
        int high = num.size()-1;

        while(low < high)
        {
            int mid1 = (low+high)/2;
            int mid2 = mid1+1;
            if(num[mid1] < num[mid2])
                low = mid2;
            else
                high = mid1;
        }
        return low;
    }
};

File: 101-200/161-170/162-findPeakElement/findPeakElement.py
class Solution(object):
    def check_peak(self, nums, i, length):
        left = right = True
        if i < length - 1:
            right = nums[i] > nums[i+1]
        if i > 0:
            left = nums[i] > nums[i-1]

        return left * 2 + right

    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        length = len(nums)
        left, right = 0, length - 1

        while left <= right:
            mid = left + (right - left) / 2
            peak = self.check_peak(nums, mid, length)
            if peak == 3:
                return mid
            elif peak == 2:
                left = mid + 1
            else:
                right = mid - 1

        return -1
